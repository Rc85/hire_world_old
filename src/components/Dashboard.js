import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import '../styles/Dashboard.css';
import Response from './Response';
import EditUser from './EditUser';

class Dashboard extends Component {
    render() {
        let component;

        if (this.props.user) {
            component = <section id='dashboard' className='main-panel'>
                <section className='tab-bar'>
                    <div className='tab-button active'>Edit</div>
                    <div className='tab-button'>My Listings</div>
                    <div className='tab-button'>Offers</div>
                    <div className='tab-button'>Accepted</div>
                    <div className='tab-button'>Settings</div>
                </section>

                <EditUser />
            </section>;
        } else {
            component = <Response code={401} header='Unauthorized Access' message={`You're not logged in`} />;
        }
        
        return(
            <div>{component}</div>
        )
    }
}

const mapStateToProps = state => {
    return {
        user: state.Login.user
    }
}

export default withRouter(connect(mapStateToProps)(Dashboard));