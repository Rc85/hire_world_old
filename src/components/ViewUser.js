import React, { Component } from 'react';
import '../styles/ViewUser.css';
import UserProfilePic from './UserProfilePic';
import ViewUserContacts from './ViewUserContacts';
import ViewUserProfile from './ViewUserProfile';
import UserReview from './UserReview';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

class ViewUser extends Component {
    render() {
        let component;

        if (this.props.user) {
            component = <div id='view-user'>
                <div className='blue-panel deep rounded'>
                    <div className='row'>
                        <div className='col-3'>
                            <UserProfilePic url={this.props.user.avatar_url} editable={false}/>

                            <ViewUserContacts />
                        </div>

                        <div className='col-9'>
                            <ViewUserProfile />
                        </div>
                    </div>
                </div>

                <UserReview />
            </div>
        }

        return(
            <div>{component}</div>
        )
    }
}

const mapStateToProps = state => {
    return {
        user: state.Login.user
    }
}

export default withRouter(connect(mapStateToProps)(ViewUser));