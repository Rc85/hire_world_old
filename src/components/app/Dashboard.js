import React, { Component } from 'react';
import { withRouter, NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import SideBar from '../includes/site/SideBar';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCommentAlt, faCog, faSyncAlt, faBriefcase, faLink } from '@fortawesome/pro-solid-svg-icons';
import BottomBar from '../includes/site/BottomBar';
import { connect } from 'react-redux';
import Response from './Response';
import { faFilePlus, faUser } from '@fortawesome/pro-solid-svg-icons';
import fetch from 'axios';
import { PromptOpen, PromptReset } from '../../actions/PromptActions';

class Dashboard extends Component {
    constructor(props) {
        super(props);
    }
    
    componentDidUpdate(prevProps, prevState) {
        if (prevProps.location.key !== this.props.location.key) {
            fetch.post('/api/get/announcements')
            .then(resp => {
                if (resp.data.status === 'success') {
                    let ids = [];
                    let localIds = JSON.parse(localStorage.getItem('dismissed'));

                    for (let a of resp.data.announcements) {
                        if (!localIds) {
                            ids.push(a.announcement_id);
                        } else if (localIds && localIds.indexOf(a.announcement_id) < 0) {
                            ids.push(a.announcement_id);
                        }
                    }

                    this.setState({announcements: resp.data.announcements, announcementIds: ids});
                }
            })
            .catch(err => LogError(err, '/api/get/announcements'));
        }
    }

    componentDidMount() {
        fetch.post('/api/get/announcements')
        .then(resp => {
            if (resp.data.status === 'success') {
                let ids = [];
                let localIds = JSON.parse(localStorage.getItem('dismissed'));

                for (let a of resp.data.announcements) {
                    if (!localIds) {
                        ids.push(a.announcement_id);
                    } else if (localIds && localIds.indexOf(a.announcement_id) < 0) {
                        ids.push(a.announcement_id);
                    }
                }

                this.setState({announcements: resp.data.announcements, announcementIds: ids});
            }
        })
        .catch(err => LogError(err, '/api/get/announcements'));
    }
    
    render() {
        let dashboard;

        let items = [
            {name: 'My Profile', link: '/dashboard/profile', active: /^\/dashboard\/profile/.test(this.props.location.pathname), icon: <FontAwesomeIcon icon={faUser} />},
            {name: 'Conversations', link: '/dashboard/conversations', active: /^\/dashboard\/conversations/.test(this.props.location.pathname), icon: <FontAwesomeIcon icon={faCommentAlt} />, messageCount: parseInt(this.props.user.messages)},
            {name: 'Post a Job', link: '/dashboard/post/job', active: this.props.location.pathname === '/dashboard/post/job', icon: <FontAwesomeIcon icon={faFilePlus} />},
            {name: 'My Jobs', link: '/dashboard/jobs', active: /^\/dashboard\/(jobs|job|posted|applied|saved)/.test(this.props.location.pathname), icon: <FontAwesomeIcon icon={faBriefcase} />, messageCount: parseInt(this.props.user.proposals) + parseInt(this.props.user.job_messages.opened_job_message_count) + parseInt(this.props.user.job_messages.active_job_message_count), items: [
                {name: 'Saved', active: this.props.location.pathname === '/dashboard/saved/jobs', link: '/dashboard/saved/jobs'},
                {name: 'Posted', active: /^\/dashboard\/posted\/(jobs|job)/.test(this.props.location.pathname), link: '/dashboard/posted/jobs'},
                {name: 'Applied', active: this.props.location.pathname === '/dashboard/applied/jobs', link: '/dashboard/applied/jobs'},
                {name: 'Proposals', active: this.props.location.pathname === '/dashboard/jobs/opened', link: '/dashboard/jobs/opened', messageCount: parseInt(this.props.user.proposals) + parseInt(this.props.user.job_messages.opened_job_message_count)},
                {name: 'Active', active: this.props.location.pathname === '/dashboard/jobs/active', link: '/dashboard/jobs/active', messageCount: parseInt(this.props.user.job_messages.active_job_message_count)},
                {name: 'Completed', active: this.props.location.pathname === '/dashboard/jobs/complete', link: '/dashboard/jobs/complete'},
                {name: 'Abandoned', active: this.props.location.pathname === '/dashboard/jobs/abandoned', link: '/dashboard/jobs/abandoned'}
            ]},
            {name: 'Link Work', link: '/dashboard/link_work', active: this.props.location.pathname === '/dashboard/link_work', icon: <FontAwesomeIcon icon={faLink} />},
            {name: 'Subscription', link: '/dashboard/subscription', active: this.props.location.pathname === '/dashboard/subscription', icon: <FontAwesomeIcon icon={faSyncAlt} />},
            {name: 'Settings', link: '/dashboard/settings/account', active: /^\/dashboard\/settings/.test(this.props.location.pathname), icon: <FontAwesomeIcon icon={faCog} />, items: [
                {name: 'Account', active: this.props.location.pathname === '/dashboard/settings/account', link: '/dashboard/settings/account'},
                {name: 'Payment', active: this.props.location.pathname === '/dashboard/settings/payment', link: '/dashboard/settings/payment'},
                {name: 'Link Work', active: this.props.location.pathname === '/dashboard/settings/link_work', link: '/dashboard/settings/link_work'}
            ]}
        ];

        if (this.props.user.status === 'activation required') {
            dashboard = <Response code={403} header={'Forbidden'} message='You need to activate your account'><div><NavLink to='/resend'>Resend Confirmation Email</NavLink></div></Response>;
        } else {
            dashboard = <React.Fragment>
                {this.props.location.pathname.match(/^\/dashboard\/edit$/) ? <div id='main-panel-bg'></div> : ''}      
                {this.props.children}
            </React.Fragment>;
        }

        return(
            <section id='dashboard'>
                {!this.props.config.mobile ? <SideBar user={this.props.user} items={items} /> : <BottomBar user={this.props.user} items={items} />}

                <div id='dashboard-main'>
                    <div className={`dashboard-main-overlay ${this.props.menu.id === 'browse-menu' ? 'show' : ''}`} />
                    {dashboard}
                </div>
            </section>
        );
    }
}

Dashboard.propTypes = {
    user: PropTypes.object.isRequired,
}

const mapStateToProps = state => {
    return {
        config: state.Config,
        user: state.Login,
        prompt: state.Prompt,
        menu: state.Menu
    }
}

export default withRouter(connect(mapStateToProps)(Dashboard));