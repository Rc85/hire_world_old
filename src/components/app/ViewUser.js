import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ViewUserContacts from '../includes/page/ViewUserContacts';
import ViewUserProfile from '../includes/page/ViewUserProfile';
import { withRouter, Redirect } from 'react-router-dom';
import fetch from 'axios';
import Loading from '../utils/Loading';
import ViewUserReviews from '../includes/page/ViewUserReviews';
import { Alert } from '../../actions/AlertActions';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faUserCircle, faEye, faExclamationTriangle, faHeart, faCoins, faUserPlus, faUserMinus, faBan, faUser } from '@fortawesome/pro-solid-svg-icons';
import ViewUserBusinessHours from '../includes/page/ViewUserBusinessHours';
import { connect } from 'react-redux';
import { LogError } from '../utils/LogError';
import MessageSender from '../includes/page/MessageSender';
import { faCheckCircle, faTimesCircle } from '@fortawesome/pro-regular-svg-icons';
import TitledContainer from '../utils/TitledContainer';
import Tooltip from '../utils/Tooltip';
import { ShowConfirmation, ResetConfirmation } from '../../actions/ConfirmationActions';
import JobProposal from '../includes/page/JobProposal';
import ViewUserWorkHistory from '../includes/page/ViewUserWorkHistory';
import { ShowLoading, HideLoading } from '../../actions/LoadingActions';
import { ShowSelectionModal, ResetSelection, ResetSelectionConfirm } from '../../actions/SelectModalActions';

class ViewUser extends Component {
    constructor(props) {
        super(props);

        this.state = {
            user: {},
            status: 'Loading',
            hours: {},
            userReported: false,
            isFriend: false,
            isBlocked: false,
            jobs: [],
            message: '',
            stats: {
                view_count: 0,
                job_complete: 0,
                job_abandon: 0,
                last_login: null
            }
        }
    }
    
    componentDidUpdate(prevProps) {
        if (prevProps.location.key !== this.props.location.key) {
            this.setState({status: 'Loading'});

            fetch.post('/api/get/user', {username: this.props.match.params.username, id: this.props.match.params.listing_id})
            .then(resp => {
                if (resp.data.status === 'success') {
                    this.setState({user: resp.data.user, stats: resp.data.stats, hours: resp.data.hours, status: '', userReported: resp.data.userReported, isFriend: resp.data.isFriend, jobs: resp.data.jobs, isBlocked: resp.data.isBlocked});
                } else if (resp.data.status === 'error') {
                    this.setState({status: ''});
                    this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
                } else if (resp.data.status === 'access error') {
                    this.setState({status: resp.data.status, statusMessage: resp.data.statusMessage});
                }
            })
            .catch(err => LogError(err, '/api/get/user'));
        }

        if (this.props.selection.data.action) {
            if (!prevProps.selection.selected && !this.props.selection.selected && this.props.selection.confirm) {
                this.props.dispatch(Alert('error', 'Please select a reason'));
                this.props.dispatch(ResetSelectionConfirm());
            } else if (this.props.selection.confirm !== prevProps.selection.confirm) {
                if (this.props.selection.data.action === 'report user' && this.props.selection.selected && this.props.selection.confirm) {
                    this.submitReport(this.props.selection.selected, this.props.selection.specified);
                    this.props.dispatch(ResetSelection());
                }
            }
        }
    }
    
    componentDidMount() {
        fetch.post('/api/get/user', {username: this.props.match.params.username, id: this.props.match.params.listing_id, url: this.props.location.pathname})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({user: resp.data.user, stats: resp.data.stats, hours: resp.data.hours, status: '', userReported: resp.data.userReported, isFriend: resp.data.isFriend, jobs: resp.data.jobs, isBlocked: resp.data.isBlocked});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
                this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
            } else if (resp.data.status === 'access error') {
                this.setState({status: resp.data.status, statusMessage: resp.data.statusMessage});
            }
        })
        .catch(err => {
            LogError(err, '/api/get/user')
        });
    }

    submitReport(reason, specified) {
        this.props.dispatch(ShowLoading('Sending report'));

        fetch.post('/api/report/submit', {url: this.props.location.pathname, type: 'User Profile', reason: reason, specified: specified === '' ? null : specified})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({userReported: true});
            }

            this.props.dispatch(HideLoading());
            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => {
            LogError(err, '/api/report/submit');
            this.setState({status: ''});
            this.props.dispatch(Alert('error', 'An error occurred'));
        });
    }

    sendMessage(message, verified, subject) {
        this.setState({sendStatus: 'Sending'});

        fetch.post('/api/conversation/submit', {subject: subject, message: message, user: this.state.user.username, verified: verified})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({status: '', sendStatus: 'send success', message: ''});
            } else if (resp.data.status === 'send error' || resp.data.status === 'error') {
                this.setState({status: '', sendStatus: resp.data.status});
            }

            setTimeout(() => {
                this.setState({sendStatus: ''});
            }, 5000);

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => {
            this.setState({status: '', sendStatus: 'error'});
            this.props.dispatch(Alert('error', 'An error occurred'));
            LogError(err, '/api/conversation/submit');
        });
    }

    friendUser(action) {
        this.setState({status: 'Adding'});

        fetch.post('/api/user/friend', {user: this.state.user.username, action: action})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({status: '', isFriend: action === 'add'});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
            }

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        }).catch(err => {
            this.setState({status: ''});
            this.props.dispatch(Alert('error', 'An error occurred'));
            LogError(err, '/api/user/friend');
        });
    }

    blockUser(action) {
        this.setState({status: 'Blocking'});

        fetch.post('/api/user/block', {user: this.state.user.username, action: action})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({status: '', isBlocked: action === 'block'});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
            }

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => {
            this.setState({status: ''});
            LogError(err, '/api/user/block');
        });
    }

    submitProposal(data) {
        this.setState({status: 'Submitting Proposal'});

        fetch.post('/api/job/create', {...data, user: this.state.user.username})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({status: '', message: ''});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
            }

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => {
            LogError(err, '/api/job/create');
            this.setState({status: ''});
        });
    }

    render() {
        let contacts, profile, reportButton, message, friendIcon, businessHours, jobs, blockIcon;

        if (this.state.status === 'access error') {
            return <Redirect to={`/error/listing/404`} />
        } else if (this.state.status === 'Loading') {
            return <Loading size='7x' />;
        } else if (this.state.status === 'redirect') {
            return <Redirect to='/main' />;
        }

        if (this.state.user) {
            contacts = <ViewUserContacts user={this.state.user} />;
            profile = <ViewUserProfile user={this.state.user} stats={this.state.stats} hours={this.state.hours} />;

            if (this.state.user.display_business_hours) {
                businessHours = <ViewUserBusinessHours hours={this.state.hours} />;
            }

            if (this.props.user.user && this.props.user.user.username !== this.state.user.username) {
                if (!this.state.isFriend) {
                    friendIcon = <Tooltip text='Add to Friends List' placement='bottom-right'><FontAwesomeIcon icon={faUserPlus} className='view-button text-alt-highlight mr-2' onClick={() => this.friendUser('add')} /></Tooltip>;
                } else {
                    friendIcon = <Tooltip text='Remove from Friends List' placement='bottom-right'><FontAwesomeIcon icon={faUserMinus} className='view-button text-alt-highlight mr-2' onClick={() => this.friendUser('remove')} /></Tooltip>
                }

                if (!this.state.isBlocked) {
                    blockIcon = <Tooltip text='Block User' placement='bottom-right'><FontAwesomeIcon icon={faTimesCircle} className='view-button text-dark mr-2' onClick={() => this.blockUser('block')} /></Tooltip>
                } else {
                    blockIcon = <Tooltip text='Unblock User' placement='bottom-right'><FontAwesomeIcon icon={faTimesCircle} className='view-button text-danger mr-2' onClick={() => this.blockUser('unblock')} /></Tooltip>
                }
                
                if (!this.state.userReported) {
                    reportButton = <Tooltip text='Report this listing' placement='bottom-right'><FontAwesomeIcon icon={faExclamationTriangle} className='view-button text-warning' onClick={() => this.props.dispatch(ShowSelectionModal(`Please select a reason`, ['Spam', 'Suspicious', 'Duplicate', 'Impersonating', 'Inappropriate'], {action: 'report user'}))} /></Tooltip>;
                } else {
                    reportButton = <Tooltip text='Reported' placement='bottom-right'><FontAwesomeIcon icon={faExclamationTriangle} className='text-dark' /></Tooltip>;
                }

                if (this.state.message === 'message') {
                    if (this.state.user.username !== this.props.user.user.username) {
                        message = <React.Fragment>
                            <hr/>

                            <MessageSender send={(message, verified, subject) => this.sendMessage(message, verified, subject)} status={this.state.sendStatus} className='mt-4' cancel={() => this.setState({message: ''})} withSubject />
                        </React.Fragment>
                        ;
                    }
                } else if (this.state.message === 'propose a job') {
                    message = <React.Fragment>
                        <hr/>
                        
                        <JobProposal className='mt-4' cancel={() => this.setState({message: ''})} submit={(data) => this.submitProposal(data)} status={this.state.status} />
                    </React.Fragment>;
                }
            }
        }
        
        return(
            <div id='view-user' className='main-panel'>
                <div id='view-user-details-container'>
                    <div id='view-user-main'>
                        {<TitledContainer title={this.state.user ? this.state.user.username : ''} bgColor='yellow' icon={<FontAwesomeIcon icon={faUserCircle} />} shadow>
                            {profile}

                            {this.props.user.user && this.state.user && this.props.user.user.username !== this.state.user.username ?
                                <div className='text-right'>
                                    {this.state.message != 'message' && this.state.user.allow_messaging ? <button className='btn btn-primary' onClick={() => this.setState({message: 'message'})}>Message</button> : ''}
                                    {this.state.message != 'propose a job' && this.state.user && this.state.user.link_work_acct_status === 'Approved' && new Date(this.state.user.subscription_end_date) > new Date() ? <button className='btn btn-success' onClick={() => this.setState({message: 'propose a job'})}>Start A Job</button> : ''}
                                </div>
                            : ''}

                            {message}

                            <hr/>
                            
                            <div id='view-user-footer'>
                                <div id='view-user-stats'>
                                    <div className='mr-5'>
                                        <Tooltip text='Job completed' placement='bottom'>
                                            <FontAwesomeIcon icon={faCheckCircle} className='text-success mr-2' />
                                            <span>{this.state.stats.job_complete}</span>
                                        </Tooltip>
                                    </div>
                        
                                    <div className='mr-5'>
                                        <Tooltip text='Job abandoned' placement='bottom'>
                                            <FontAwesomeIcon icon={faBan} className='text-danger mr-2' />
                                            <span>{this.state.stats.job_abandon}</span>
                                        </Tooltip>
                                    </div>
        
                                    <div className='mr-5'>
                                        <Tooltip text='Views' placement='bottom'>
                                            <FontAwesomeIcon icon={faEye} className='mr-2' />
                                            <span>{this.state.stats.view_count}</span>
                                        </Tooltip>
                                    </div>
                                </div>
        
                                <div className='view-buttons'>
                                    {friendIcon}
                                    {blockIcon}
                                    {reportButton}
                                </div>
                            </div>
                        </TitledContainer>}
                    </div>

                    <div id='view-user-details'>
                        {contacts}
                        {businessHours}
                        <ViewUserWorkHistory user={this.state.user.username} />
                    </div>

                    <ViewUserReviews username={this.props.match.params.username} user={this.props.user} listingId={this.props.match.params.listing_id} />
                </div>
            </div>
        )
    }
}

ViewUser.propTypes = {
    user: PropTypes.object
}

const mapStateToProps = state => {
    return {
        user: state.Login,
        confirm: state.Confirmation,
        selection: state.Selection
    }
}

export default withRouter(connect(mapStateToProps)(ViewUser));