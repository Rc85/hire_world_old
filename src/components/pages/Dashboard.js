import React from 'react';
import { withRouter } from 'react-router-dom';
import Response from './Response';
import TabBar from '../utils/TabBar';
import Loading from '../utils/Loading';
import PropTypes from 'prop-types';

const Dashboard = props => {
    let success = /success$/;
    let fail = /fail$/;

    if (success.test(props.user.status) || props.user.user) {
        console.log(props.user);

        return(
            <section id='dashboard' className='main-panel w-100'>
                <TabBar items={[
                    {name: 'Edit', active: props.location.pathname === '/dashboard/edit', link: '/dashboard/edit'},
                    {name: 'List', active: props.location.pathname === '/dashboard/list', link: '/dashboard/list'},
                    {name: 'Inquiries', active: /^\/dashboard\/message(s)?\/Inquire.*/.test(props.location.pathname), link: '/dashboard/messages/Inquire'},
                    {name: 'Active Jobs', active: /^\/dashboard\/message(s)?\/Active.*/.test(props.location.pathname), link: '/dashboard/messages/Active'},
                    {name: 'Completed', active: /^\/dashboard\/message(s)?\/Complete.*/.test(props.location.pathname), link: '/dashboard/messages/Complete'},
                    {name: 'Abandoned', active: /^\/dashboard\/message(s)?\/Abandoned.*/.test(props.location.pathname), link: '/dashboard/messages/Abandoned'},
                    {name: 'Settings', active: props.location.pathname === '/dashboard/settings', link: '/dashboard/settings'},
                ]} />

                {props.children}
            </section>
        )
    } else if (fail.test(props.user.status)) {
        return(
            <Response code={401} header='Unauthorized Access' message={`You're not logged in`} />
        )
    } else if (!props.user.status) {
        return(
            <Loading size='7x' />
        )
    } else if (!props.user.user) {
        return(
            <Response code={500} header='Internal Server Error' message={'An unknown error has occurred. Please contact the administrator.'} />
        )
    } else {
        return(
            <div></div>
        )
    }
}

Dashboard.propTypes = {
    user: PropTypes.object.isRequired
}

export default withRouter(Dashboard);