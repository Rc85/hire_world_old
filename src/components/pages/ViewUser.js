import React, { Component } from 'react';
import ViewUserSocialMedia from '../includes/page/ViewUserSocialMedia';
import ViewUserContacts from '../includes/page/ViewUserContacts';
import ViewUserProfile from '../includes/page/ViewUserProfile';
import { withRouter, Redirect } from 'react-router-dom';
import fetch from 'axios';
import Loading from '../utils/Loading';
import Response from './Response';
import ViewUserStats from '../includes/page/ViewUserStats';
import ViewUserReview from '../includes/page/ViewUserReview';
import SubmitReview from '../includes/page/SubmitReview';
import { Alert } from '../../actions/AlertActions';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBuilding } from '@fortawesome/free-regular-svg-icons';
import { faUserCircle, faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
import { connect } from 'react-redux';
import { UncontrolledTooltip } from 'reactstrap';
import { LogError } from '../utils/LogError';
import MessageSender from '../includes/page/MessageSender';

class ViewUser extends Component {
    constructor(props) {
        super(props);

        this.state = {
            user: null,
            reviews: [],
            status: '',
            submitReview: false,
            review: '',
            hours: {},
            reportedReviews: [],
            userReported: false
        }
    }

    componentDidUpdate(prevProps) {
        if (prevProps.location.key !== this.props.location.key) {
            this.setState({status: 'Loading'});

            fetch.post('/api/get/user', {username: this.props.match.params.username})
            .then(resp => {
                if (resp.data.status === 'success') {
                    this.setState({user: resp.data.user, reviews: resp.data.reviews, stats: resp.data.stats, hours: resp.data.hours, status: '', reportedReviews: resp.data.reports, userReported: resp.data.userReported});
                } else if (resp.data.status === 'error') {
                    this.setState({status: ''});
                    this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
                } else if (resp.data.status === 'error page') {
                    this.setState({status: resp.data.status, statusMessage: resp.data.statusMessage});
                }
            })
            .catch(err => LogError(err, '/api/get/user'));
        }
    }
    
    componentDidMount() {
        this.setState({status: 'Loading'});

        fetch.post('/api/get/user', {username: this.props.match.params.username})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({user: resp.data.user, reviews: resp.data.reviews, stats: resp.data.stats, hours: resp.data.hours, status: '', reportedReviews: resp.data.reports, userReported: resp.data.userReported});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
                this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
            } else if (resp.data.status === 'error page') {
                this.setState({status: resp.data.status, statusMessage: resp.data.statusMessage});
            }
        })
        .catch(err => LogError(err, '/api/get/user'));
    }

    submitReview(review, star) {
        let blankCheck = /^\s*$/;

        if (blankCheck.test(review)) {
            this.props.dispatch(Alert('error', 'Review cannot be blank'));
        } else {
            this.setState({status: 'Sending'});

            fetch.post('/api/review/submit', {review: review, star: star, reviewing: this.state.user.username})
            .then(resp => {
                if (resp.data.status === 'success') {
                    let reviews = this.state.reviews;

                    if (resp.data.review) {
                        reviews.unshift(resp.data.review);
                    }

                    this.setState({status: '', reviews: reviews, submitReview: false});
                } else if (resp.data.status === 'error') {
                    this.setState({status: ''});
                }
            
                this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
            })
            .catch(err => LogError(err, '/api/review/submit'));
        }
    }

    editReview(review, index) {
        let reviews = this.state.reviews;
        reviews[index] = review;

        this.setState({reviews: reviews});
    }

    submitReport() {
        this.setState({status: 'Sending'});

        fetch.post('/api/report/submit', {id: this.state.user.user_id, type: 'User', url: this.props.location.pathname, user: this.state.user.username})
        .then(resp => {
            if (resp.data.status === 'success') {
                this.setState({status: '', userReported: true});
            } else if (resp.data.status === 'error') {
                this.setState({status: ''});
            } else if (resp.data.status === 'redirect') {
                this.setState({status: resp.data.status});
            }

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => LogError(err, '/api/report/submit'));
    }

    sendMessage(message, subject) {
        this.setState({status: 'Sending'});

        fetch.post('/api/message/submit', {subject: subject, message: message, user: this.state.user})
        .then(resp => {
            console.log(resp);
            this.setState({status: '', sendStatus: resp.data.status});

            this.props.dispatch(Alert(resp.data.status, resp.data.statusMessage));
        })
        .catch(err => LogError(err, '/api/message/submit'));
    }

    render() {
        let status, contacts, socialMedia, profile, reviews, submitReview, submitReviewButton, reviewed, reportButton, businessName, message;

        if (this.state.reviews && this.props.user) {
            reviewed = this.state.reviews.findIndex(review => review.reviewer === this.props.user.username)
        }

        if (this.state.status === 'Sending') {
            status = <Loading size='5x' />;
        }

        if (this.state.user) {
            contacts = <ViewUserContacts user={this.state.user} />;
            socialMedia = <ViewUserSocialMedia user={this.state.user} listing={this.state.user ? {id: this.state.user.listing_id, status: this.state.user.listing_status} : {}} />;
            profile = <ViewUserProfile user={this.state.user} />;

            if (this.state.user.user_business_name) {
                businessName = <div className='d-flex-center mb-3'><div className='w-10 mr-1'><FontAwesomeIcon icon={faBuilding} className='view-user-icon mr-1' size='lg' /></div> {this.state.user.user_website ? <a href={this.state.user.user_website}>{this.state.user.user_business_name}</a> : this.state.user.user_business_name}</div>;
            }

            if (this.state.reviews.length > 0) {
                reviews = this.state.reviews.map((review, i) => {
                    let reported = false;

                    if (this.state.reportedReviews.indexOf(review.review_id) >= 0) reported = true;

                    return <ViewUserReview key={i} review={review} user={this.props.user} edit={(review) => this.editReview(review, i)} reported={reported} />
                });
            } else {
                reviews = <div className='text-center mt-5'>
                    <h1 className='text-muted'>Not Yet Reviewed</h1>
                </div>
            }

            if (this.state.submitReview) {
                submitReview = <SubmitReview submit={(review, star) => this.submitReview(review, star)} cancel={() => this.setState({submitReview: false })} />;
            } else if (!this.state.submitReview && this.props.user && this.state.user) {
                if (this.props.user.username !== this.state.user.username && reviewed < 0) {
                    submitReviewButton = <button className='btn btn-primary' onClick={() => this.setState({submitReview: true})}>Submit Review</button>;
                }
            }

            if (this.state.status === 'error page') {
                return <Response header={'Error'} message={this.state.statusMessage} />;
            } else if (this.state.status === 'Loading') {
                return <Loading size='7x' />;
            } else if (this.state.status === 'redirect') {
                return <Redirect to='/account/login' />;
            }

            if (this.props.user && this.props.user.username !== this.state.user.username) {
                if (!this.state.userReported) {
                    reportButton = <span>
                        <FontAwesomeIcon icon={faExclamationTriangle} size='xs' className='menu-button' id='report-user-button' onClick={() => this.submitReport()} />
                        <UncontrolledTooltip placement='top' target='report-user-button'>Report this user</UncontrolledTooltip>
                    </span>;
                } else {
                    reportButton = <span>
                        <FontAwesomeIcon icon={faExclamationTriangle} size='xs' className='theme-bg' id='report-user-button' />
                        <UncontrolledTooltip placement='top' target='report-user-button'>Already reported</UncontrolledTooltip>    
                    </span>;
                }

                if (this.state.user.username !== this.props.user.username) {
                    message = <MessageSender send={(message, subject) => this.sendMessage(message, subject)} status={this.state.sendStatus} />;
                }
            }
        }
        
        return(
            <div id='view-user' className='main-panel w-100'>
                {status}
                <div className='blue-panel shallow rounded'>
                    <div className='row'>
                        <div className='col-3'>
                            {businessName}
                            {contacts}
                            <ViewUserStats stats={this.state.stats || {}} hours={this.state.hours} />
                            {socialMedia}
                            <div className='text-right'>{reportButton}</div>
                        </div>

                        <div className='col-9'>
                            {profile}

                            {message}
                        </div>
                    </div>
                </div>

                <div className='text-right mt-3'>
                    <div className='w-75 mx-auto'>{submitReviewButton}</div>

                    <div className='mt-1'>
                        {submitReview}
                    </div>
                </div>

                <div id='user-reviews' className='mt-5'>
                    {reviews}
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        user: state.Login.user
    }
}

export default withRouter(connect(mapStateToProps)(ViewUser));