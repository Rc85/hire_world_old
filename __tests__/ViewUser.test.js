import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router';
import { reducers } from '../src/reducers';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import Dashboard from '../src/components/pages/Dashboard';
import ViewUser from '../src/components/pages/ViewUser';
import puppeteer from 'puppeteer';
import toJson from 'enzyme-to-json';

const store = createStore(reducers, applyMiddleware(thunk));
const user = {user: true, status: 'success'}

Enzyme.configure({adapter: new Adapter()});

describe('Messages page', () => {
    test('snapshot', async() => {
        const wrapper = shallow(
            <Provider store={store}>
                <MemoryRouter initialEntries={[{pathname: '/user/some-user', key: 'view-user'}]}>
                    <Dashboard user={user}><ViewUser user={user} /></Dashboard>
                </MemoryRouter>
            </Provider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});