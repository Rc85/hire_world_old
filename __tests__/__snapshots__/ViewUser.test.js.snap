// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages page snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "view-user",
        "pathname": "/user/some-user",
      },
    ]
  }
>
  <withRouter(Dashboard)
    user={
      Object {
        "status": "success",
        "user": true,
      }
    }
  >
    <withRouter(Connect(ViewUser))
      user={
        Object {
          "status": "success",
          "user": true,
        }
      }
    />
  </withRouter(Dashboard)>
</MemoryRouter>
`;
